compositor_pipeline/src/queue/audio_queue.rs

```diff
@@ -0,0 +1,163 @@
+use std::{
+    cmp::min,
+    collections::{HashMap, VecDeque},
+    time::Duration,
+};
+
+use compositor_render::{AudioSamples, AudioSamplesBatch, AudioSamplesSet, InputId};
+
+use crate::pipeline::AudioChannels;
+
+use super::QueueError;
+
+pub struct AudioQueue {
+    input_queues: HashMap<InputId, InputQueue>,
+    timestamp_offsets: HashMap<InputId, Duration>,
+}
+
+impl AudioQueue {
+    pub fn new() -> Self {
+        AudioQueue {
+            input_queues: HashMap::new(),
+            timestamp_offsets: HashMap::new(),
+        }
+    }
+
+    pub fn add_input(&mut self, input_id: InputId, channels: AudioChannels, sample_rate: u32) {
+        let input_queue = InputQueue::new(channels, sample_rate);
+        self.input_queues.insert(input_id, input_queue);
+    }
+
+    pub fn remove_input(&mut self, input_id: &InputId) {
+        self.input_queues.remove(input_id);
+    }
+
+    pub fn enqueue_samples(
+        &mut self,
+        input_id: InputId,
+        samples_batch: AudioSamplesBatch,
+    ) -> Result<(), QueueError> {
+        let Some(input_queue) = self.input_queues.get_mut(&input_id) else {
+            return Err(QueueError::UnknownInputId(input_id));
+        };
+        let offset = *self
+            .timestamp_offsets
+            .entry(input_id)
+            .or_insert_with(|| samples_batch.pts);
+
+        input_queue.enqueue_samples(samples_batch.samples, samples_batch.pts - offset)?;
```

Comment:
If I understand correctly, this causes every audio stream's first buffer to have `pts == 0`. I think this is incorrect, shouldn't the inserted offset be `samples_batch.pts - current_queue_pts`, or something similar?